packages/coin-sol/src/utils/transactionUtil.ts

```diff
@@ -0,0 +1,164 @@
+    this.numberReadonlySignedAccount = rawTx.numberReadonlySignedAccount;
+    this.numberReadonlyUnSignedAccount = rawTx.numberReadonlyUnSignedAccount;
+    this.keyCount = rawTx.keyCount;
+    this.keys = rawTx.keys;
+    this.recentBlockHash = rawTx.recentBlockHash;
+    this.instructionCount = rawTx.instructionCount;
+    this.programIdIndex = rawTx.programIdIndex;
+    this.keyIndicesCount = rawTx.keyIndicesCount;
+    this.keyIndices = rawTx.keyIndices;
+    this.dataLength = rawTx.dataLength;
+    this.data = rawTx.data;
+  }
+
+  dataEncode(amount: number | string, decimals?: number | undefined): string {
+    const programIdToNumber = Number(this.programIdIndex);
+    const isNormalTransfer = programIdToNumber === 2;
+    const programIdIndexSpan = isNormalTransfer ? 4 : 1;
+    const dataAlloc = isNormalTransfer ? 12 : 9;
+    const data = Buffer.alloc(dataAlloc);
+
+    data.writeUIntLE(programIdToNumber, 0, programIdIndexSpan);
+
+    if (isNormalTransfer) {
+      const v2e32 = Math.pow(2, 32);
+      const value = Number(amount) * params.LAMPORTS_PER_SOL;
+      const hi32 = Math.floor(value / v2e32);
+      const lo32 = value - hi32 * v2e32;
+
+      data.writeUInt32LE(lo32, programIdIndexSpan);
+      data.writeInt32LE(hi32, programIdIndexSpan + 4);
+    } else {
+      const LAMPORTS_PER_TOKEN = new BN(10).pow(new BN(decimals));
+      const value = new BN(amount).mul(LAMPORTS_PER_TOKEN);
+      const valueHex = value.toString(16, 8 * 2);
+      const valueBuf = Buffer.from(valueHex, 'hex').reverse();
+
+      data.write(valueBuf.toString('hex'), programIdIndexSpan, 8, 'hex');
+    }
+    this.data = data.toString('hex');
+    return data.toString('hex');
+  }
+
+  serialize(): string {
+    return (
+      this.sigCount +
+      this.signature +
+      this.numberRequireSignature +
+      this.numberReadonlySignedAccount +
+      this.numberReadonlyUnSignedAccount +
+      this.keyCount +
+      this.keys +
+      this.recentBlockHash +
+      this.instructionCount +
+      this.programIdIndex +
+      this.keyIndicesCount +
+      this.keyIndices +
+      this.dataLength +
+      this.data
+    );
+  }
+  serializeArgument(): string {
+    return this.keys + this.recentBlockHash + this.dataLength + this.data;
+  }
+}
+
+export function formTransaction(transaction: TransactionType, txType: string, signature?: string): RawTransaction {
+  const { fromPubkey, toPubkey, recentBlockhash, amount, options } = transaction;
+  const { programId, data, owner, decimals, value } = options as TransactionOptions;
+
+  const keys = addressToHex(owner) + addressToHex(fromPubkey) + addressToHex(toPubkey) + addressToHex(programId);
+  const recentBH = base58.decode(recentBlockhash).toString('hex');
+
+  const rawTx = new RawTransaction({
+    signature,
+    numberRequireSignature: '01',
+    numberReadonlySignedAccount: '00',
+    numberReadonlyUnSignedAccount: '01',
+    keyCount: '03',
```

Comment:
Verifying a transaction with the same `fromPubkey` and same `toPubkey` might not succeed.
Seems that `rawTx` is not the same.

Expected result:

```javascript
import * as web3 from "@solana/web3.js";

const FROM_PUBKEY = new web3.PublicKey(
  "9rzi118z4WUui9Hm6GbE61FyAfdF23HNBvsUXWJYpAS1"
);

const transaction = new web3.Transaction({
  recentBlockhash: "cRJSiaHpCa5CXEr5xyAuw8avnP1aKvqycFCh8sSAq3m",
  feePayer: FROM_PUBKEY,
}).add(
  web3.SystemProgram.transfer({
    fromPubkey: FROM_PUBKEY,
    toPubkey: FROM_PUBKEY,
    lamports: web3.LAMPORTS_PER_SOL * 0.1,
  })
);

console.log(transaction.compileMessage().serialize().toString("hex"));

```

Reply:
This issue seems to cause by  https://github.com/solana-labs/solana-web3.js/blob/d93efdfe8dd51a91e254a800be63c9157a9fe7cc/src/transaction.ts#L273-L285.

Reply:
Indeed rawTx of this case is not the same, but in actual scenario do user actually need to run into this case. I mean they would also have to pay for gas fee and not really doing anything.

Reply:
Currently, there might not have this kind of scenario. But I want to encode the transaction more flexibly.
It is hard to promise that kind of scenario will not happen in the future.

Reply:
Ok I got it

Reply:
And also spl-token `rawTx` is not the same as the rawTx generated by `web3.js`.

```javascript
import * as web3 from "@solana/web3.js";
import { createTransferInstruction } from "@solana/spl-token";

/** Address of the SPL Token program */
const TOKEN_PROGRAM_ID = new web3.PublicKey(
  "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
);

/** Address of the SPL Associated Token Account program */
const ASSOCIATED_TOKEN_PROGRAM_ID = new web3.PublicKey(
  "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
);

const FROM_PUBKEY = new web3.PublicKey(
  "9rzi118z4WUui9Hm6GbE61FyAfdF23HNBvsUXWJYpAS1"
);

const TO_PUBKEY = new web3.PublicKey(
  "AnJZ8PLH3YZVJRifPb2jLXDwaKXtScHrQmpCiQ3vS8jm"
);

const USDT_TOKEN = new web3.PublicKey(
  "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"
);

(async () => {
  const [FROM_ADDRESS] = await web3.PublicKey.findProgramAddress(
    [
      FROM_PUBKEY.toBuffer(),
      TOKEN_PROGRAM_ID.toBuffer(),
      USDT_TOKEN.toBuffer(),
    ],
    ASSOCIATED_TOKEN_PROGRAM_ID
  );

  const [TO_ADDRESS] = await web3.PublicKey.findProgramAddress(
    [TO_PUBKEY.toBuffer(), TOKEN_PROGRAM_ID.toBuffer(), USDT_TOKEN.toBuffer()],
    ASSOCIATED_TOKEN_PROGRAM_ID
  );

  const splTokenTransaction = new web3.Transaction({
    recentBlockhash: "cRJSiaHpCa5CXEr5xyAuw8avnP1aKvqycFCh8sSAq3m",
    feePayer: FROM_PUBKEY,
  }).add(
    createTransferInstruction(
      FROM_PUBKEY,
      FROM_ADDRESS,
      TO_ADDRESS,
      web3.LAMPORTS_PER_SOL * 0.1
    )
  );

  console.log(
    "SPL Token transaction: ",
    splTokenTransaction.compileMessage().serialize().toString("hex")
  );
})();
```

Reply:
Seem like if amount is float dataEncode function would return incorrect data, but even if tx data was correct we still can't sign this tx due to unknown error, could you check it out pls

Reply:
Please trying to upgrade CoolWalletPro with [coolwallet-signing-tools](https://github.com/CoolBitX-Technology/coolwallet-signing-tools) and try it again, Thanks!